### ISSUE TITLE
Missing Android N compatibility because of behavior changes

### ISSUE THREAD

**Issue by TheNephilim88** [sqlcipher/android-database-sqlcipher#216]
Hello,

Google changed some things regarding linking/loading libs: https://developer.android.com/preview/behavior-changes.html#ndk

This also affects SQLCipher. E.g. the Test-Suite failed 10 out of 32 tests.
![screenshot_20160310-091222](https://cloud.githubusercontent.com/assets/1174584/13663240/7124f7a4-e6a0-11e5-957c-3b06bcdd4dd2.png)

Log show up following:
```
03-10 09:09:17.497 26856-26869/net.zetetic:provider W/linker: library "libutils.so" ("/system/lib/libutils.so") needed or dlopened by "/data/app/net.zetetic-1/lib/arm/libsqlcipher_android.so" is not accessible for the namespace "classloader-namespace" - the access is temporarily granted as a workaround for http://b/26394120
03-10 09:09:17.497 26856-26869/net.zetetic:provider W/linker: library "libcutils.so" ("/system/lib/libcutils.so") needed or dlopened by "/data/app/net.zetetic-1/lib/arm/libsqlcipher_android.so" is not accessible for the namespace "classloader-namespace" - the access is temporarily granted as a workaround for http://b/26394120
03-10 09:09:17.517 26856-26869/net.zetetic:provider W/linker: library "libnativehelper.so" ("/system/lib/libnativehelper.so") needed or dlopened by "/data/app/net.zetetic-1/lib/arm/libdatabase_sqlcipher.so" is not accessible for the namespace "classloader-namespace" - the access is temporarily granted as a workaround for http://b/26394120
03-10 09:09:17.518 26856-26869/net.zetetic:provider W/linker: library "libandroid_runtime.so" ("/system/lib/libandroid_runtime.so") needed or dlopened by "/data/app/net.zetetic-1/lib/arm/libdatabase_sqlcipher.so" is not accessible for the namespace "classloader-namespace" - the access is temporarily granted as a workaround for http://b/26394120
03-10 09:09:17.518 26856-26869/net.zetetic:provider W/linker: library "libbinder.so" ("/system/lib/libbinder.so") needed or dlopened by "/data/app/net.zetetic-1/lib/arm/libdatabase_sqlcipher.so" is not accessible for the namespace "classloader-namespace" - the access is temporarily granted as a workaround for http://b/26394120
```

**Comment by developernotes** [comment]
Hello @TheNephilim88 

Thank you for your early report of SQLCipher for Android on Android N.  We hope to begin looking into the Android N build soon.  Take care!

**Comment by staticfish** [comment]
As per https://developer.android.com/preview/behavior-changes.html#ndk 

>Use of SSL_ctrl symbol from libcrypto.so should be replaced with an app local version. For example, you should statically link libcyrpto.a in your .so file or include your own dynamically libcrypto.so from BoringSSL or OpenSSL in your app.

So instead of OpenSSL being compiled for each arch and being statically linked in the .so, I believe now that the bordingSSL library so (or a?) should be linked directly through the ndk from the user's OS. OpenSSL has been deprecated and removed in favour of boringssl

**Comment by developernotes** [comment]
Hi @staticfish 

We have historically built a static OpenSSL because of the differing versions of OpenSSL bundled across the various Android OS's.  We will consider this change in the future.

**Comment by TheNephilim88** [comment]
Did anyone found time to look into this?
I know Android N is still dev preview, but its annoying nonetheless, especially if users dont understand and are giving bad ratings because of it :(

**Comment by sjlombardo** [comment]
The changes required to address this problem are not trivial and will likely require some major modifications to the library. We are currently in the middle of release preparation for an unrelated update. Once that is complete we will switch over to work on this issue. 

I understand that this problem is an inconvenience, however, as you mentioned N is still an early developer release.

**Comment by neutron666** [comment]
But that's what dev previews are for, make issues appear early and fix them. I really would like to continune testing the N dev preview, but i need a app based on this lib urgently and it's very annoying to being able to use it. PLEASE start on fixing this problem as soon as possible.

**Comment by mikecousins** [comment]
Hopefully this gets prioritized soon. The ease for user's to install the latest dev preview has resulted in quite a few people not being able to run anything using your library.

**Comment by smarkovik** [comment]
Hello guys, 
any ideas when this will get in your pipeline?
N preview has been out for a while and the apps which depend on sql cipher are essentially blocked.

**Comment by sjlombardo** [comment]
@smarkovik we just finished up a release of SQLCipher yesterday. Now that is out of the way, we'll be switching over to look at this issue as a priority.

**Comment by smarkovik** [comment]
Hello @sjlombardo, 
Not being pushy but just curious of your progress? any ETAs?
could the community help ?

**Comment by developernotes** [comment]
Hi @smarkovik 

We are actively working on this project and will make our changes available as soon as everything is working.  Thanks!

**Comment by mvojjala** [comment]
Hi @sjlombardo 

We are also in same situation. Not able use our app on Android N because of this issue. Just want to check with you if you have any update or tentative dates of a possible update. 

Thanks

**Comment by sjlombardo** [comment]
@mvojjala we're making appreciable progress on this issue, but it does involve some fairly major changes. Keep an eye on this ticket, and/or at https://discuss.zetetic.net/c/sqlcipher for information about upcoming availability for testing.

**Comment by fuentespatrick** [comment]
I appreciate the effort and the update @sjlombardo.
My team is similarly blocked, so the significant work is greatly appreciated!

**Comment by developernotes** [comment]
Hi folks,

We have just announced a beta of SQLCipher for Android with support for Android Developer N Preview [here](https://discuss.zetetic.net/t/sqlcipher-for-android-upcoming-changes-for-android-n-and-coordinated-beta-test-request/1315). Please take a look if you are interested in helping test out these changes. Thanks!

**Comment by smarkovik** [comment]
Awsome work guys !

**Comment by fuentespatrick** [comment]
Thanks, we'll give it a try!

**Comment by zokipirlo** [comment]
Tnx. I think it will work, but I have also a dependency on IOCipher. It's crashing while loading `stlport_shared` inside static block in `VirtualFileSystem.java`. I tried to remove it, but it's still loading it in `iocipher`. Is there any easy solution to make it work?

**Comment by developernotes** [comment]
Hi @zokipirlo 

It sounds as if IOCipher may need to be rebuilt referencing the new version of SQLCipher for Android. IOCipher has publish build instructions on their README [here](https://github.com/guardianproject/IOCipher/blob/master/README.md).

**Comment by zokipirlo** [comment]
It works great. Everything looks fine. Thank you.
I was having some problems with building IOCipher but that's another story :) Doesn't matter.

**Comment by developernotes** [comment]
Hello @zokipirlo 

Thanks for the report, we are glad to hear it is working well for you!

**Comment by brodybits** [comment]
I am also really happy that you guys made the solution. I am especially happy that you guys removed the ICU and native setLocale parts, though it may be nice to have a version with ICU for those who may need it. I encounter 3 minor failures when I ran the SQLCipher test suite:
- Unicode (ICU) test
- Closed database test
- Invalid password test

The following change would the invalid password test:
```diff
--- a/src/main/java/net/zetetic/ZeteticApplication.java
+++ b/src/main/java/net/zetetic/ZeteticApplication.java
@@ -53,7 +53,10 @@ public class ZeteticApplication extends Application {
     public SQLiteDatabase createDatabase(File databaseFile){
         Log.i(TAG, "Entered ZeteticApplication::createDatabase");
         Log.i(TAG, "Before SQLiteDatabase.openOrCreateDatabase");
-        return SQLiteDatabase.openOrCreateDatabase(databaseFile, DATABASE_PASSWORD, null);
+        SQLiteDatabase database = SQLiteDatabase.openOrCreateDatabase(databaseFile, DATABASE_PASSWORD, null);
+        Log.i(TAG, "Populating the database");
+        database.rawExecSQL("CREATE TABLE IF NOT EXISTS MyTest (TestData)");
+        return database;
     }
 
     public void extractAssetToDatabaseDirectory(String fileName) throws IOException {
```

The closed database test fails because it tries the SQLiteDatabase.setLocale function which then throws an java.lang.UnsatisfiedLinkError with a "Native method not found" message. Considering the native_setLocale method is no longer implemented, I would favor fixing SQLiteDatabase.setLocale to just throw a runtime exception with a message that the method is no longer implemented.

**Comment by developernotes** [comment]
Hi @brodybits 

Thank you for your feedback, it is much appreciated! Have you pulled down the latest changes to the SQLCipher for Android [test suite](https://github.com/sqlcipher/sqlcipher-android-tests)? It should address those tests and also include a few new tests too.

**Comment by brodybits** [comment]
> Have you pulled down the latest changes to the SQLCipher for Android test suite?

I missed it, I definitely like the changes to the test suite. I will give it a try sometime next week and let you know if I find anything else.

An extra point is that I wish you guys would consider supporting a version that something like BoringSSL which supports 64-bit builds out of the box.

**Comment by developernotes** [comment]
>I missed it, I definitely like the changes to the test suite. I will give it a try sometime next week and let you know if I find anything else.

Sounds good.

>An extra point is that I wish you guys would consider supporting a version that something like BoringSSL which supports 64-bit builds out of the box.

The 1.1.0 release of OpenSSL, [currently scheduled](https://www.openssl.org/policies/releasestrat.html) for release on May 12th, 2016 will include support for x64.

**Comment by brodybits** [comment]
I just ran the test suite and it seems to run faster than before. Assuming that OpenSSL 1.1.0 is available next week will you release this change first or wait to include OpenSSL 1.1.0? (I would favor waiting for OpenSSL 1.1.0.)

**Comment by developernotes** [comment]
Hi @brodybits 

Yes, we would like to include the OpenSSL 1.1.0 release in a future beta build. We would like to gather ample feedback from the community regarding these changes before we integrate them forward as a public release.

**Comment by zokipirlo** [comment]
Internal testing looks fine. I released an app update on Play Store with this version of Sqlcipher :) Will tell you soon if any crash reports will occur.

**Comment by brodybits** [comment]
I also find it pretty strange that I still cannot build this on Android API 23 and raised PR #230 to fix this.

**Comment by brodybits** [comment]
In the past I could do `make init` and it would automatically use the installed NDK build implementation. In the Android N preview I get a message that I have to define `ANDROID_NDK_ROOT` when building (I only have to define this for the `make init` step). I hope this will be documented.

Sorry to be a pain about this but I find it strange that I cannot build this with `android-23` and raised #231 which I think fixes the build the right way. If not, I hope we can find another solution.

**Comment by developernotes** [comment]
>In the past I could do make init and it would automatically use the installed NDK build implementation. In the Android N preview I get a message that I have to define ANDROID_NDK_ROOT when building (I only have to define this for the make init step).

The is due to building the OpenSSL libraries from the [`make init`](https://github.com/sqlcipher/android-database-sqlcipher/blob/android-n-preview/Makefile#L19-L26) step. Previously we would build them locally and check them into source control. The aim with this branch was to streamline some of those efforts, so building the OpenSSL dependency once within `make init` was practical. The requirement to define `ANDROID_NDK_ROOT` was always present in the `build-openssl-libraries.sh` file; [here](https://github.com/sqlcipher/android-database-sqlcipher/blob/master/build-openssl-libraries.sh#L4-L7) is the relevant portion in the file found in the master branch.

**Comment by brodybits** [comment]
Ah now I understand. I hope this will be documented.

**Comment by developernotes** [comment]
Hi @brodybits 

The documentation has been updated. Thanks

**Comment by mvojjala** [comment]
Hi @sjlombardo 

Everything works great with the beta libraries in our app except for logs and I see that it already reported. Thank you for your effort.
Any time line on when will be the final version released ?

Thanks

**Comment by sjlombardo** [comment]
@mvojjala thanks for letting us know things are working well for you. We haven't finalized a time frame yet, as we are still collecting feedback on the beta. We will likely release one or two more beta versions that will reduce the logging, and if the upcoming OpenSSL release works as planned incorporates a new version. Please keep an eye out, we'll be sure to announce upcoming releases on the site, and will make a note on this ticket as well.

**Comment by marcardar** [comment]
Is there an AAR package out there for those of us on slow connections...

**Comment by brodybits** [comment]
FYI `make clean` doesn't seem to work for me.

**Comment by brodybits** [comment]
While I am not in a position to participate in the official beta testing I have integrated a version with the `android-n-preview` changes in my Cordova plugin. I am using a custom version at https://github.com/litehelpers/android-database-sqlcipher-api-fix with the following additional fixes:

- Fixes to support API 23 as proposed in #231
- Working `distclean` target

**Comment by developernotes** [comment]
Hello @marcardar,

>Is there an AAR package out there for those of us on slow connections...

No, if you prefer to not build from source you are welcome to request the beta builds through the support channel mentioned in [this](https://discuss.zetetic.net/t/sqlcipher-for-android-upcoming-changes-for-android-n-and-coordinated-beta-test-request/1315) post, however they are distributed as a zip, not AAR. Keep in mind, cloning and initializing the `android-n-preview` branch is much smaller, and compiles much faster than the `master` branch.

**Comment by developernotes** [comment]
Hi @brodybits 

>FYI make clean doesn't seem to work for me.

What issue are you having with `make clean`?

**Comment by brodybits** [comment]
This is interesting: if I build the project successfully (or just finish `make init` successfully) then do `make clean` it seems to be OK. But if I stop `make init` in the middle or `make init` fails then do `make clean` I get the following (cut off after the make error):

```
cd /Users/brodybits/cordova-plugin/android-database-sqlcipher-api-fix/jni && \
	ndk-build clean
Android NDK: ERROR:/Users/brodybits/cordova-plugin/android-database-sqlcipher-api-fix/jni/Android.mk:static-libcrypto: LOCAL_SRC_FILES points to a missing file    
Android NDK: Check that /Users/brodybits/cordova-plugin/android-database-sqlcipher-api-fix/jni/../external/android-libs/armeabi/libcrypto.a exists  or that its path is correct   
/usr/local/Cellar/android-ndk/r11c/build/core/prebuilt-library.mk:45: *** Android NDK: Aborting    .  Stop.
make: [clean-ndk] Error 2 (ignored)
```

I would be happy if we just add a working `distclean` target that blows away all generated artifacts. For example: https://github.com/litehelpers/android-database-sqlcipher-api-fix/commit/b0fa9d3c8d7d0229fd8fc7424cdac7d8f9521e6e (please let me know if you want me to issue a PR)

**Comment by marcardar** [comment]
Thanks @developernotes . Unfortunately, I don't have an Android N device for testing so don't satisfy the requirements for obtaining the zip. I do have several pre-N devices though and satisfy all other items, so it's a shame you have that requirement.

Thanks very much for the suggestion to only clone that branch. I'm trying that now, so hopefully my connection can manage that!

**Comment by developernotes** [comment]
Hello @marcardar 

Thank you for your interest, I've reached out to you directly.

**Comment by neerajgautam** [comment]
I am trying to launch my app on Android N Preview 3 , but getting following error on launching and app crashes. Can anyone please help!!  I am stuck. Thanks in advance.

05-20 15:22:53.589: E/linker(8449): library "/system/lib/libjavacore.so" ("/system/lib/libjavacore.so") needed or dlopened by "/APP_PACKAGE/lib/arm/SOFILE NAME.so" is not accessible for the namespace: [name="classloader-namespace", ld_library_paths="", default_library_paths="/APP_PACKAGE/lib/arm:/system/fake-libs:/APP_PACKAGE/base.apk!/lib/armeabi", permitted_paths="/data:/mnt/expand:/data/data/APP_PACKAGE"]

**Comment by powder366** [comment]
Please fix.

**Comment by lena8912** [comment]
The phone I am using is 6p, I got error as follows:
java.lang.UnsatisfiedLinkError: dalvik.system.PathClassLoader[DexPathList[[zip file "/data/app/com.google.android.apps.talk-1/base.apk"],nativeLibraryDirectories=[/data/app/com.google.android.apps.talk-1/lib/arm64, /data/app/com.google.android.apps.talk-1/base.apk!/lib/arm64-v8a, /vendor/lib64, /system/lib64]]] couldn't find "libsqlcipher.so"

**Comment by brodybits** [comment]
> The phone I am using is 6p, I got error as follows:
> java.lang.UnsatisfiedLinkError: dalvik.system.PathClassLoader[DexPathList[[zip file "/data/app/com.google.android.apps.talk-1/base.apk"],nativeLibraryDirectories=[/data/app/com.google.android.apps.talk-1/lib/arm64, /data/app/com.google.android.apps.talk-1/base.apk!/lib/arm64-v8a, /vendor/lib64, /system/lib64]]] couldn't find "libsqlcipher.so"

@lena8912 your app has one or more 64-bit libraries but this project does not include a 64-bit library which is why you get this error. Please be sure to remove **all** 64-bit libraries from your project's `libs` and from any dependencies that may be included.

**Comment by developernotes** [comment]
Hi @neerajgautam,

That error message doesn't appear to reference SQLCipher for Android. Did you have SQLCipher for Android previously integrated within your application?

**Comment by Pratyul** [comment]
Do we have an rough estimate on dates when the N support with be GA. It will help everyone in planning their releases.

**Comment by zokipirlo** [comment]
Just release it :)
I did it almost a month ago and didn't receive a single crash.
Even `net.sqlcipher.database.SQLiteException: not an error: Could not register Android SQL functions.` is gone from logs.

**Comment by developernotes** [comment]
Hi @Pratyul 

We don't have a firm date set at this point, however we are looking at the near term to allow for developers to transition and test their applications prior to Android N becoming release to manufacturing. Have you had an opportunity to try out the `android-n-preview` branch?

**Comment by Pratyul** [comment]
@developernotes 
Have tried the android-n-preview branch on Android N and even previous OS versions. Have not seen any major issues or crashes so far. We wanted to plan Android N release and sql cipher would be an important part of that. 
Even an approximate date like 1 month, 2 months will be good enough for the plan.

**Comment by lena8912** [comment]
I am using this code-branch on Android 5.1.1 and encountered a crash cannot reproduce.
`                         --------- beginning of crash
06-02 14:29:07.107 A/libc: Fatal signal 11 (SIGSEGV), code 2, fault addr 0xb7e43330 in tid 28485 (JDWP)

06-02 14:29:24.558 W/linker: libsqlcipher.so: unused DT entry: type 0x6ffffffe arg 0x1f5f8

06-02 14:29:24.558 W/linker: libsqlcipher.so: unused DT entry: type 0x6fffffff arg 0x3`
maybe someone saw this before? really hope some could help me with this, it took me the whole afternoon without any chance to figure it out..

**Comment by developernotes** [comment]
Hello @lena8912 

Is it correct that you only received this error once, you haven't been able to reproduce it again? What device did you experience this on, or was it on an emulator?

**Comment by lena8912** [comment]
Yes I receive this error once and not able to reproduce it again. I'm using lg nexus 4, it's a real device and running 5.1.1 and I used android-n-preview branch code.

**Comment by developernotes** [comment]
Hello @lena8912 

If you find you are able to reproduce this again, it would be beneficial to build the binaries locally and symbolicate the native stack trace from the adb logs. You can do this by executing the following before you run your application:

```
adb logcat | $NDK/ndk-stack -sym $PROJECT_PATH/obj/local/armeabi
```

The above `$NDK` references the root directory of your NDK folder and the `$PROJECT_PATH` is the root folder of the SQLCipher for Android source tree (following a build). The above is also relative to utilization of the `armeabi` binaries, adjust according to your target platform.

**Comment by lena8912** [comment]
Hello @developernotes 
I haven't reproduced the crash yet, but I followed your instruction anyway.
I got the following hint using : 
```
$ adb logcat | $ANDROID_NDK_ROOT/ndk-stack -sym /obj/local/armeabi

********** Crash dump: **********
Build fingerprint: 'google/occam/mako:5.1/LMY47O/1783956:user/release-keys'
pid: 28475, tid: 28485, name: JDWP  >>> [packagename here] <<<
signal 11 (SIGSEGV), code 2 (SEGV_ACCERR), fault addr 0xb7e43330
Stack frame #00 pc 000a1330  [heap]
Stack frame #01 pc 00137ba1  /system/lib/libart.so (art::gc::Heap::GetObjectsAllocated() const+28)
Stack frame #02 pc 000f08bd  /system/lib/libart.so (art::Dbg::DdmSendHeapInfo(art::Dbg::HpifWhen)+912)
Stack frame #03 pc 000f0c83  /system/lib/libart.so (art::Dbg::DdmHandleHpifChunk(art::Dbg::HpifWhen)+114)
Stack frame #04 pc 0001a5e9  /data/dalvik-cache/arm/system@framework@boot.oat
Crash dump is completed

********** Crash dump: **********
Build fingerprint: 'google/occam/mako:5.1/LMY47O/1783956:user/release-keys'
pid: 28672, tid: 28682, name: JDWP  >>> [packagename here] <<<
signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr 0x550004
Stack frame #00 pc 00137b9e  /system/lib/libart.so (art::gc::Heap::GetObjectsAllocated() const+25)
Stack frame #01 pc 000f08bd  /system/lib/libart.so (art::Dbg::DdmSendHeapInfo(art::Dbg::HpifWhen)+912)
Stack frame #02 pc 000f0c83  /system/lib/libart.so (art::Dbg::DdmHandleHpifChunk(art::Dbg::HpifWhen)+114)
Stack frame #03 pc 0001a5e9  /data/dalvik-cache/arm/system@framework@boot.oat
Crash dump is completed

********** Crash dump: **********
Build fingerprint: 'google/occam/mako:5.1/LMY47O/1783956:user/release-keys'
pid: 1573, tid: 1583, name: JDWP  >>> [packagename here] <<<
signal 11 (SIGSEGV), code 2 (SEGV_ACCERR), fault addr 0xb7d9d2a0
Stack frame #00 pc 000002a0  [heap]
Stack frame #01 pc 00137ba1  /system/lib/libart.so (art::gc::Heap::GetObjectsAllocated() const+28)
Stack frame #02 pc 000f08bd  /system/lib/libart.so (art::Dbg::DdmSendHeapInfo(art::Dbg::HpifWhen)+912)
Stack frame #03 pc 000f0c83  /system/lib/libart.so (art::Dbg::DdmHandleHpifChunk(art::Dbg::HpifWhen)+114)
Stack frame #04 pc 0001a5e9  /data/dalvik-cache/arm/system@framework@boot.oat
```

since the first crash dump's : `pid: 28475, tid: 28485, name: JDWP`
is the same with my crash's tid = 28485:
                         ```
 --------- beginning of crash
06-02 14:29:07.107 A/libc: Fatal signal 11 (SIGSEGV), code 2, fault addr 0xb7e43330 in tid 28485 (JDWP)
```
Is this information useful? Thanks for ur reply!!

**Comment by developernotes** [comment]
Hi @lena8912 

Thanks for looking into this further. The crash appears to be sourced within libc, and your log above references a linker warning for libsqlcipher.so. Do you still see the linker warnings even if there is no crash? Thanks

**Comment by lena8912** [comment]
@developernotes 
Yes, there is a log whether crash or not. It appears every time I install and run my app(so I think it is not the reason why my app crashed). see below:
```
06-02 14:29:24.558 W/linker: libsqlcipher.so: unused DT entry: type 0x6ffffffe arg 0x1f5f8
06-02 14:29:24.558 W/linker: libsqlcipher.so: unused DT entry: type 0x6fffffff arg 0x3`
```
it is shown in yellow color in logcat, so I think it might be just a warning.
Any clues?!

**Comment by jaredpetker** [comment]
Building and usage on the x86 emulator works for me, however, building onto my Nexus 5x on newest Android M seems to produce this:

```
 java.lang.UnsatisfiedLinkError: dlopen failed: cannot locate symbol "__aeabi_memcpy4" referenced by "/data/app/tigerconnect.app-2/lib/arm/libsqlcipher.so"...
                                                                      at java.lang.Runtime.loadLibrary(Runtime.java:372)
                                                                      at java.lang.System.loadLibrary(System.java:1076)
                                                                      at net.sqlcipher.database.SQLiteDatabase.loadLibs(SourceFile:177)
                                                                      at net.sqlcipher.database.SQLiteDatabase.loadLibs(SourceFile:170)
```

Issues with loading the sqlcipher.so it seems, any reason it can't find the symbol that I'm missing?

**Comment by developernotes** [comment]
Hi @jaredpetker 

What branch of SQLCipher for Android are you compiling? What is the version number of your Android NDK? Are you running Android "M" (i.e., Marshmallow), or Android "N", the current beta OS on your Nexus 5X?

**Comment by jaredpetker** [comment]
@developernotes So it seems like the issue has to do with compiling sqlcipher with the newest (r12) ndk, using r11c seems to work fine (across x86 and arm for me). Seems to be an issue only for arm devices as my x86 emulator worked fine with the sqlcipher compiled with r12.

**Comment by developernotes** [comment]
Hi @jaredpetker 

Thanks for the quick response, I'm glad to hear that the issue was due to the NDK toolchain (unfortunately that is). Our [building](https://github.com/sqlcipher/android-database-sqlcipher/tree/android-n-preview#building) instructions suggest using r11c, glad to hear that is working well for you, too!

**Comment by jaredpetker** [comment]
Yea, I saw that initially and decided to live on the wild side which sadly bit me back, I assume you all have some pipeline internally for worrying about building with newer toolchains, until then yea, seems like all is well thanks!

**Comment by ashishmody** [comment]
Hey @developernotes, do we know when will the update with this fix be released? Do we have a timeline in place? Looks like Google froze the APIs for N and is encouraging developers to get their apps ready for N soon. Any timeline would really help us plan! Thanks!

**Comment by smarkovik** [comment]
I second that!

On Tue, Jun 21, 2016, 08:10 ashishmody <notifications@github.com> wrote:

> Hey @developernotes <https://github.com/developernotes>, do we know when
> will the update with this fix be released? Do we have a timeline in place?
> Looks like Google froze the APIs for N and is encouraging developers to get
> their apps ready for N soon. Any timeline would really help us plan! Thanks!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/sqlcipher/android-database-sqlcipher/issues/216#issuecomment-227349278>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AA8OPaFtx3G2bo27ZlI2z4Dkipaasx3lks5qN4A7gaJpZM4Htgon>
> .
>
-- 

Tancho

**Comment by developernotes** [comment]
Hi @ashishmody, @smarkovik 

We are planning a release soon, now that the N API has been finalized, and we've tested the library against the latest Developer Preview 4 (i.e., NPD56N). We appreciate your patience and enthusiasm!

**Comment by developernotes** [comment]
Hi folks,

This is just an update, we have just release SQLCipher for Android 3.5.0 with support for Android N. More information can be found [here](http://discuss.zetetic.net/t/sqlcipher-for-android-release-android-n-support/1465). Thanks!

**Comment by dfyx** [comment]
Will there be official binaries? The link in https://www.zetetic.net/sqlcipher/sqlcipher-for-android/ doesn't work for anything newer than 3.1.0 and when I build my own binaries, they are much larger (5 MB for armeabi and armeabi-v7a and 7 MB for x86) than the 1.6 MB mentioned in the blog post.

**Comment by dfyx** [comment]
Nevermind, found the aar at http://central.maven.org/maven2/net/zetetic/android-database-sqlcipher/3.5.0/android-database-sqlcipher-3.5.0.aar which is just a renamed zip file.

(I just need the .so files for a .NET project)

**Comment by developernotes** [comment]
Hi @dfyx 

The native libraries included in the AAR package are compiled for the Android OS specifically and will not run on another environment.

**Comment by dfyx** [comment]
@developernotes I know. I was talking about a Xamarin Android project.

**Comment by liuzhihuinb** [comment]
@developernotes
when the column is defined by TEXT, It seems that the cursor.getInt() method can't return the right value of the column.

**Comment by developernotes** [comment]
Hi @liuzhihuinb 

I was not able to reproduce the problem you describe with the SQLCipher for Android [test suite](https://github.com/sqlcipher/sqlcipher-android-tests). Are you doing something different?

**Comment by liuzhihuinb** [comment]
@developernotes 
Thank you for you reply.
I found the problem in my app when I updated the sqlCihper to the version of 3.5.1.
As my app worked perfectly with the sqlCihper before,I'm confused.
I have wrote a demo for using cursor.getInt(0) ,I think you can reproduce with the code below.
https://github.com/liuzhihuinb/SqlCipherDemo.git

**Comment by developernotes** [comment]
Hi @liuzhihuinb 

It appears your application may have a problem with the `min` function instead? Are you able to reproduce the issue with the SQLCipher for Android [test suite](https://github.com/sqlcipher/sqlcipher-android-tests)? We have a [test](https://github.com/sqlcipher/sqlcipher-android-tests/blob/master/src/main/java/net/zetetic/tests/TextAsIntegerTest.java) that uses `TEXT` as the storage type, but requests an integer back from the cursor correctly..

**Comment by liuzhihuinb** [comment]
Hi @developernotes 
Yes, I had reproduced this problem in the SQLCipher for android test suite.
As this issue had been closed. I submitted a new issue below since other people may encounter the same issue. 

https://github.com/sqlcipher/android-database-sqlcipher/issues/244
